### ISSUE TITLE
tests not passing

### ISSUE THREAD

**Issue by bd4** [argonne-lcf/SyclCPLX#31]
The tests are not passing for me. I've tried different oneAPI versions, different gpu and cpu backends, different subsets fail/pass. Is there a known combination that everything passes on, what was the primary test environment when the Catch2 pr was merged? @AidanBeltonS @jle-quel 

I would like to have CI that runs the testsuite before merge, this actually works with OpenCL CPU driver from Intel on latest github action runners.

@TApplencourt says there is a known issue with how shared memory and Catch2 interact, so switching to device mem with explicit migration should fix some of the issues.

**Comment by jle-quel** [comment]
I was running it on `opencl::gpu` and `level0::gpu`

I've been working on the `marray` today, updating the implementation, proposal and test.

I did see some complex tests failing when running on CPU, which I fixed as part of my `marray` work; these were all due to two reasons:
- some kernels are not wrapped under `if device handles aspect`, which fails due to the non-support for `sycl::half`
- the check function is called when the kernel cannot be run due to the aspect, and thus the result is incorrect.

I will test again the full complex tomorrow (opencl::cpu and opencl::gpu) and will push a quick patch tomorrow.

Finally, I don't remember having any errors; I did test yesterday again the llvm implementation with the llvm-test-suite and everything was fine.

Finally, locally when I created the PR, I did not have any errors and the same for when I tested it with the llvm-test suite.

**Comment by bd4** [comment]
Ok thanks! I will wait on making any further changes to the tests. Trying to get things in order before the paper is due, so we can address any issues that come up. Once the tests are fixed, I can rebase the CI PR and hopefully it will work to make sure tests stay working going forward.

Are you running in integrated GPU, or a discrete GPU on test hardware?

**Comment by TApplencourt** [comment]
>Ok thanks! I will wait on making any further changes to the tests. Trying to get things in order before the paper is due, so we can address any issues that come up. Once the tests are fixed, I can rebase the CI PR and hopefully it will work to make sure tests stay working going forward.

Np! 

>Are you running in integrated GPU, or a discrete GPU on test hardware?

I think only CPU are available on public CI

**Comment by TApplencourt** [comment]
> Finally, locally when I created the PR, I did not have any errors and the same for when I tested it with the llvm-test suite.

Make sense. The catch2  issue only show up on PVC. Should be fine with Gen9 / other integrated hardware.

(delete my previous answers I was not reply to the correct comment)

**Comment by jle-quel** [comment]
I created a PR https://github.com/argonne-lcf/SyclCPLX/pull/32 to update the implementations to fix some errors I saw with the llvm-test-suite and also updated the tests.

As you mentioned, I, too, saw the failures when running on the CPU, and that's what I saw yesterday when I was working on the `marray`.

I also had a failure with `asin`, which I remember having already when I started working on the complex codebase.
I remember investigating it, and the result was correct, even though the test failed. The reason
`sin(asin(x)) == x`; but due to precision, `x` was usually `~ x.002`. I augmented the tolerance multiplier for this specific case.

I ran the test (with the new PR) using multiple backend end devices (`level_zero:gpu`, `opencl:gpu` and `opencl:cpu`), and they were all passing.

However, I did have some time, flaky results (SIGABRT) with the assign operators.

**Comment by TApplencourt** [comment]
> I also had a failure with asin, which I remember having already when I started working on the complex codebase.
I remember investigating it, and the result was correct, even though the test failed. The reason
sin(asin(x)) == x; but due to precision, x was usually ~ x.002. I augmented the tolerance multiplier for this specific case.

Oh yeah, I think maybe all the reciprocal will have some problem. I Was thinking that we already got a special tolerance for those... Maybe it was just 2x the "normal" one.

>However, I did have some time, flaky results (SIGABRT) with the assign operators.

Oh interesting...

**Comment by jle-quel** [comment]
> > However, I did have some time, flaky results (SIGABRT) with the assign operators.
> 
> Oh interesting...

Note that this result was already present with the previous complex implementation.
I'm leaning towards Catch2, we already saw SEGFAULT in the past, but it's not confirmed.

**Comment by bd4** [comment]
With #43, most of the tests are passing now, Catch2 + shared memory was the cause, using device mem works around it.

There are 7 tests still failing, all pow with complex exponent with nan or inf related. It looks like it might be a real bug in Inte OneAPI, clang and gcc behave differently. In particular, `std:pow` and `std::exp(y * log(x))` produce different results when build with oneapi, and are the same on upstream/clang gcc. This is without sycl being involved at all. 
```
$ ./build-clang/src/exp_complex_host 
===================================
x      (4.42,2.02)
y      (-inf,inf)
log(x) (1.58098,0.428671)
y * log(x) (-inf,-nan)
exp(y*log(x)) (0,-0)
pow    (0,-0)
===================================
x      (4.42,2.02)
y      (nan,inf)
log(x) (1.58098,0.428671)
y * log(x) (-inf,inf)
exp(y*log(x)) (0,0)
pow    (0,0)

$ ./build-oneapi/src/exp_complex_host 
===================================
x      (4.42,2.02)
y      (-inf,inf)
log(x) (1.58098,0.428671)
y * log(x) (-inf,-nan)
exp(y*log(x)) (0,0)
pow    (nan,nan)
===================================
x      (4.42,2.02)
y      (nan,inf)
log(x) (1.58098,0.428671)
y * log(x) (-inf,inf)
exp(y*log(x)) (0,0)
pow    (nan,nan)
```

**Comment by bd4** [comment]
Also worth noting, most of the tests still fail with CI. I will try updating to 2023.1.0 and latest opencl cpu drivers and see if I can get it to work, but it may be a lost cause within the virtualization environment used by github. When I run the tests on opencl:cpu they have same 7 failures as PVC, so is really specific to CI. I have different issues on laptop with Gen11 i7 cpu, so may not be able to get it working everywhere. I think latest oneAPI release + PVC + at least one CPU is a good start.
